src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs

```diff
@@ -863,6 +863,45 @@ public IEnumerable<SyntaxTree> GetSyntaxTreesWithNameFromDeclarationOnlyCompilat
                 return version;
             }
             #endregion
+
+            private class CompilationReferencesReferenceSameCompilationComparer : IEqualityComparer<MetadataReference>
+            {
+                public static readonly CompilationReferencesReferenceSameCompilationComparer Instance = new CompilationReferencesReferenceSameCompilationComparer();
+
+                public bool Equals(MetadataReference x, MetadataReference y)
+                {
+                    // this should take care of all metadata reference types except compilation reference type
+                    // except compiler references, we cache them as long as they are from same snapshot, so reference equality will work.
+                    if (object.ReferenceEquals(x, y))
+                    {
+                        return true;
+                    }
+
+                    // here we only special case compilation reference and not portable executable reference since
+                    // consumer of this make sure they share metadata reference between compilations.
+                    var compilation1 = x as CompilationReference;
+                    var compilation2 = y as CompilationReference;
+                    if (compilation1 == null || compilation2 == null)
+                    {
+                        return false;
+                    }
+
+                    // MetadataProperty implements value equality
+                    return object.ReferenceEquals(compilation1.Compilation, compilation2.Compilation) &&
+                           object.Equals(compilation1.Properties, compilation2.Properties);
+                }
+
+                public int GetHashCode(MetadataReference obj)
+                {
+                    var c = obj as CompilationReference;
+                    if (c != null)
+                    {
+                        return c.Compilation.GetHashCode();
```

Comment:
looks like you care about Compilation and Properties in .Equals, but only Compilation in GetHashCode.  Intentional?